test/async_cache_test.dart

```diff
@@ -0,0 +1,139 @@
+    expect(await cache.fetch(call), 'Called 1');
+    await cache.invalidate();
+    expect(await cache.fetch(call), 'Called 2');
+    await cache.invalidate();
+    expect(await cache.fetch(call), 'Called 3');
+  });
+
+  test('should fetch a stream via a callback', () async {
+    var callback = expectAsync0/*<Stream<String>>*/(() {
+      return new Stream.fromIterable(['1', '2', '3']);
+    });
+    expect(await cache.fetchStream(callback).toList(), ['1', '2', '3']);
+  });
+
+  test('should not fetch stream via callback when a cache exists', () async {
+    Stream<String> doNotCall() {
+      return new Stream.fromIterable(['SHOULD', 'NOT', 'SHOW', 'UP']);
+    }
+
+    await cache.fetchStream(() async* {
+      yield '1';
+      yield '2';
+      yield '3';
+    }).toList();
+    expect(await cache.fetchStream(doNotCall).toList(), ['1', '2', '3']);
+  });
+
+  test('should not fetch stream via callback when request in flight', () async {
+    // Unlike the above test, we want to verify that we don't make multiple
+    // calls if a cache is being filled currently, and instead wait for that
+    // cache to be completed.
+    var controller = new StreamController<String>();
+    Stream<String> call() => controller.stream;
+    expect(cache.fetchStream(call).toList(), completion(['1', '2', '3']));
+    controller.add('1');
+    controller.add('2');
+    await new Future.value();
+    expect(cache.fetchStream(call).toList(), completion(['1', '2', '3']));
+    controller.add('3');
+    await controller.close();
+  });
+
+  test('should fetch stream via a callback again when cache expires', () async {
+    var timesCalled = 0;
+    Stream<String> call() {
+      return new Stream<String>.fromIterable(['Called ${++timesCalled}']);
+    }
+
+    expect(await cache.fetchStream(call).toList(), ['Called 1']);
+    expect(await cache.fetchStream(call).toList(), ['Called 1'],
+        reason: 'Cache still fresh');
+
+    time = time.add(const Duration(hours: 1) - const Duration(seconds: 1));
+    expect(await cache.fetchStream(call).toList(), ['Called 1'],
+        reason: 'Cache still fresh');
+
+    time = time.add(const Duration(seconds: 1));
+    expect(await cache.fetchStream(call).toList(), ['Called 2']);
+    expect(await cache.fetchStream(call).toList(), ['Called 2'],
+        reason: 'Cache fresh again');
+
+    time = time.add(const Duration(hours: 1));
+    expect(await cache.fetchStream(call).toList(), ['Called 3']);
+  });
+
+  test('should fetch via a callback when manually invalidated', () async {
+    var timesCalled = 0;
+    Stream<String> call() {
+      return new Stream<String>.fromIterable(['Called ${++timesCalled}']);
+    }
+    expect(await cache.fetchStream(call).toList(), ['Called 1']);
+    await cache.invalidate();
+    expect(await cache.fetchStream(call).toList(), ['Called 2']);
+    await cache.invalidate();
+    expect(await cache.fetchStream(call).toList(), ['Called 3']);
+  });
+}
```

Comment:
Add tests that assert that canceling or pausing one stream doesn't affect others.